File: 백준/Bronze/2798. 블랙잭/블랙잭.cc
#include <iostream>
#include <cmath>

using namespace std;

int Blackjack (int* array, int length, int number);

int main (int argc, char* argv[]) {
    
    int size = 0, number = 0;
    cin >> size >> number;
    int* array = new int[size];
    if (array == NULL) return 0;
    
    for (int index = 0; index < size; index++) {
        
        cin >> array[index];
    }
    
    cout << Blackjack (array, size, number) << "\n";
    delete[] array;
    return 0;
}

int Blackjack (int* array, int length, int number) {
    
    if (array == NULL) throw array;
    
    int result = 0;
    for (int first = 0; first < length; first++) {
        for (int second = first + 1; second < length; second++) {
            for (int third = second + 1; third < length; third++) {
                
                int sum = array[first] + array[second] + array[third];
                if ((sum - result) > 0 && number >= sum) result = sum;
            }
        }
    }
    return result;
}

File: 백준/Bronze/2798. 블랙잭/블랙잭.rs
use std::io::{self, BufRead};

fn main() {

    let stdin = io::stdin();
    let mut iterator = stdin.lock().lines();

    let first_line = iterator.next().unwrap().unwrap();
    let mut first_line_split = first_line.split_whitespace();
    let size: usize = first_line_split.next().unwrap().parse().unwrap();
    let target: i32 = first_line_split.next().unwrap().parse().unwrap();

    let second_line = iterator.next().unwrap().unwrap();
    let source: Vec<i32> = second_line
        .split_whitespace()
        .map(|s| s.parse().unwrap())
        .collect();

    let result = solution (&source, size, target);
    println! ("{}", result);
}

fn solution (source: &Vec<i32>, length: usize, target: i32) -> i32 {

    let mut result = 0;
    
    for first in 0..length-2 {
        for second in first+1..length-1 {
            for third in second+1..length {
                let sum = source[first] + source[second] + source[third];
                if sum <= target && sum > result {

                    result = sum;
                }
            }
        }
    }

    result
}